File: CP01031.cpp
#include <bits/stdc++.h>
using namespace std;

using namespace std;

int check(vector<int> a, int n, int sum)
{
    int l, r;
    sort(a.begin(), a.end());
    int ans = 0;

    // Every iteration of loop counts triplet with
    // first element as arr[i].
    for (int i = 0; i < n - 2; i++)
    {
        // Initialize other two elements as corner elements
        // of subarray arr[j+1..k]
        int j = i + 1, k = n - 1;

        // Use Meet in the Middle concept
        while (j < k)
        {
            // If sum of current triplet is more or equal,
            // move right corner to look for smaller values
            if (a[i] + a[j] + a[k] >= sum)
                k--;

            // Else move left corner
            else
            {
                // This is important. For current i and j, there
                // can be total k-j third elements.
                ans += (k - j);
                j++;
            }
        }
    }
    return ans;
}

int main()
{
    int t;
    cin >> t;
    while (t--)
    {
        int n, K;
        cin >> n >> K;
        vector<int> a;
        for (int i = 0; i < n; ++i)
        {
            int tmp;
            cin >> tmp;
            a.push_back(tmp);
        }
        sort(a.begin(), a.end());
        cout<<check(a, n, K);
        cout << endl;
    }
    return 0;
}
/**
 * @brief https://www.geeksforgeeks.org/count-triplets-with-sum-smaller-that-a-given-value/
 * 
 */

File: CP01031.py
def countTriplets(arr, n, sum):
    arr.sort()
    ans = 0
    for i in range(0, n-2):
        j = i + 1
        k = n-1
        while(j < k):
            if (arr[i]+arr[j]+arr[k] >= sum):
                k = k-1
            else:
                ans += (k - j)
                j = j+1

    return ans

if __name__ == '__main__':
    arr = [5, 1, 3, 4, 7]
    n = len(arr)
    sum = 12
    print(countTriplets(arr, n, sum))
